In [284]:
#IMPORTING LIBRARIES

import pandas as pd
import numpy as np
import os

# 4.5: Data Consistency Checks

## IMPORTING DATAFRAMES:

In [285]:
#IMPORTING "products.csv" Dataframe
products = pd.read_csv(r'/Users/MadrigalPaez/Documents/2 - FORMACIÓN Y TRABAJO/CAREER FOUNDRY/DATA ANALYTICS/2 - Data Immersion/Achievement 4 - Python Fundamentals for Data Analysts/08-2025 Instacart Grocery Basket Analysis/02 Data/Original Data/products.csv',
    index_col=False)

In [286]:
products.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 49693 entries, 0 to 49692
Data columns (total 5 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   product_id     49693 non-null  int64  
 1   product_name   49677 non-null  object 
 2   aisle_id       49693 non-null  int64  
 3   department_id  49693 non-null  int64  
 4   prices         49693 non-null  float64
dtypes: float64(1), int64(3), object(1)
memory usage: 1.9+ MB


In [287]:
#IMPORTING “orders_wrangled.csv” Dataframe
orders = pd.read_csv(r'/Users/MadrigalPaez/Documents/2 - FORMACIÓN Y TRABAJO/CAREER FOUNDRY/DATA ANALYTICS/2 - Data Immersion/Achievement 4 - Python Fundamentals for Data Analysts/08-2025 Instacart Grocery Basket Analysis/02 Data/Prepared Data/orders_wrangled.csv',
    index_col=False)

In [288]:
orders.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3421083 entries, 0 to 3421082
Data columns (total 8 columns):
 #   Column                  Dtype  
---  ------                  -----  
 0   Unnamed: 0              int64  
 1   order_id                int64  
 2   user_id                 int64  
 3   eval_set                object 
 4   order_number            int64  
 5   order_day_of_week       int64  
 6   order_hour_of_day       int64  
 7   days_since_prior_order  float64
dtypes: float64(1), int64(6), object(1)
memory usage: 208.8+ MB


## DATA CONSISTENCY CHECKS:

In [289]:
products.describe()

product_id      aisle_id  department_id        prices
count  49693.000000  49693.000000   49693.000000  49693.000000
mean   24844.345139     67.770249      11.728433      9.994136
std    14343.717401     38.316774       5.850282    453.519686
min        1.000000      1.000000       1.000000      1.000000
25%    12423.000000     35.000000       7.000000      4.100000
50%    24845.000000     69.000000      13.000000      7.100000
75%    37265.000000    100.000000      17.000000     11.200000
max    49688.000000    134.000000      21.000000  99999.000000

In [290]:
orders.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       order_day_of_week  order_hour_of_day  days_since_prior_order  
count       3.421083e+06       3.421083e+06            3.214874e+06  
mean        2.776219e+00       1.345202e+01            1.111484e+01  
std         2.046829e+00       4.226088e+00            9.206737e+00  
min         0.000000e+00       0.000000e+00            0.000000e+00  
25%         1.000000e+00       1.000000e+01            4.000000e+00  
50%         3.000000e+00       1.300000e+01            7.000000e+00  
75%         5.000000e+00       1.600000e+01            1.500000e+01  
max         6.000000e+00       2.300000e+01            3.000000e+01

In [291]:
#Checking Mixed-Type Data in "orders" Dataframe:

for col in orders.columns.tolist():
  weird = (orders[[col]].applymap(type) != orders[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (orders[weird]) > 0:
    print (col)
    
Error /var/folders/32/qqkdf8qx2w189kqpjp4rlrv40000gn/T/ipykernel_65712/440481828.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.

In [292]:
#Checking Mixed-Type Data in "orders" Dataframe, using map instead of applymap:
inconsistent_rows = {}
for col in orders.columns:
    common_type = orders[col].map(type).mode()[0]
    weird = orders[col].map(type) != common_type
    if weird.any():
        inconsistent_rows[col] = products.loc[weird, [col]]
        print(f"{col}: {weird.sum()}")
    else:
        print(f"{col}: No inconsistencies found")

Unnamed: 0: No inconsistencies found
order_id: No inconsistencies found
user_id: No inconsistencies found
eval_set: No inconsistencies found
order_number: No inconsistencies found
order_day_of_week: No inconsistencies found
order_hour_of_day: No inconsistencies found
days_since_prior_order: No inconsistencies found


No inconsistencies have been found in the "orders" dataframe.

In [293]:
#Checking Mixed-Type Data in "products" Dataframe:
inconsistent_rows = {}
for col in products.columns:
    types = products[col].map(type)
    common_type = types.mode()[0]
    weird = types != common_type
    if weird.any():
        inconsistent_rows[col] = products.loc[weird, [col]]
        print(f"{col}: {weird.sum()}")
    else:
        print(f"{col}: No inconsistencies found")

product_id: No inconsistencies found
product_name: 16
aisle_id: No inconsistencies found
department_id: No inconsistencies found
prices: No inconsistencies found


We found 16 potential inconsistences found on the column "product_name".

In [294]:
#Checking the 16 inconsistences found in "product_name" column:
types = products['product_name'].map(type)
common_type = types.mode()[0]
weird = types != common_type

if weird.any():
    inconsistent_rows = products.loc[weird]  # full rows with errors
    print(f"product_name: {weird.sum()} inconsistencies")
    print(inconsistent_rows)
else:
    print("product_name: 0 inconsistencies")

product_name: 16 inconsistencies
       product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187  

There are also no inconsistencies in the "producst" datafame, all 16 are missing values.

In [295]:
#Finding missing values on "orders" dataframe:
orders.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
eval_set                       0
order_number                   0
order_day_of_week              0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

In [296]:
#Finding missing values on "products" dataframe:
products.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [297]:
df_nan_products = products[products['product_name'].isnull()]

In [298]:
df_nan_products

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

## ADDRESSING MISSING VALUES:

In [299]:
df_nan_products.describe()

product_id    aisle_id  department_id     prices
count     16.000000   16.000000      16.000000  16.000000
mean    6684.000000   89.937500      10.937500  13.012500
std    12836.665242   33.731229       4.639953   3.881731
min       34.000000   26.000000       1.000000   1.200000
25%      459.250000   70.750000       7.750000  12.175000
50%     2413.000000   98.500000      11.500000  13.650000
75%     3872.750000  120.000000      14.500000  14.425000
max    40440.000000  126.000000      16.000000  20.900000

In [300]:
df_nan_products.median()

product_id       2413.0
product_name        NaN
aisle_id           98.5
department_id      11.5
prices            13.65
dtype: object

In [301]:
df_nan_products

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

We could identify and address the NaN products via the "product_id" if there is any correspondence

In [302]:
#Finding the "product_name" for the "product_id" 34, 69, 116, 262, 525, 1511, 1780, 2240, 2586, 3159, 3230, 3736, 4283, 4790, 38183, 40440:

In [303]:
product_ids = [34, 69, 116, 262, 525, 1511, 1780, 2240, 2586, 3159, 3230, 3736, 4283, 4790, 38183, 40440]
selected_products_by_id = products[products['product_id'].isin(product_ids)]

In [304]:
selected_products_by_id

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

Unfortunately, there are no other products with those product ID(s) to establish a correspondence. 

In [305]:
#Since the column type is "object", we imput "Unknown Product" to substitute the NaN in the dataframe:

In [306]:
products_no_nan = products.copy()
products_no_nan['product_name'] = products_no_nan['product_name'].fillna('Unknown Product')

In [307]:
products_no_nan.isnull().sum()

product_id       0
product_name     0
aisle_id         0
department_id    0
prices           0
dtype: int64

In [308]:
products_no_nan[products_no_nan['product_name'] == 'Unknown Product']

product_id     product_name  aisle_id  department_id  prices
33             34  Unknown Product       121             14    12.2
68             69  Unknown Product        26              7    11.8
115           116  Unknown Product        93              3    10.8
261           262  Unknown Product       110             13    12.1
525           525  Unknown Product       109             11     1.2
1511         1511  Unknown Product        84             16    14.3
1780         1780  Unknown Product       126             11    12.3
2240         2240  Unknown Product        52              1    14.2
2586         2586  Unknown Product       104             13    12.4
3159         3159  Unknown Product       126             11    13.1
3230         3230  Unknown Product       120             16    14.4
3736         3736  Unknown Product        41              8    14.8
4283         4283  Unknown Product        77              7    14.4
4790         4790  Unknown Product        91             16    14.5
38187       38183  Unknown Product        39             12    20.9
40444       40440  Unknown Product       120             16    14.8

## DUPLICATES:

In [309]:
#Finding duplicates on "products" dataframe:
products_duplicates = products_no_nan[products_no_nan.duplicated()]

In [310]:
products_duplicates

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [311]:
#Deleting duplicated from "products" dataframe:

In [312]:
products_wrangled = products_no_nan.drop_duplicates()

## CHECKING DATA INTEGRITY FOR A FINAL "products_wrangled" DATAFRAME:

In [313]:
#Checking missing values:
products_wrangled.isnull().sum()

product_id       0
product_name     0
aisle_id         0
department_id    0
prices           0
dtype: int64

In [314]:
#Checking duplicated:
products_duplicates = products_wrangled[products_wrangled.duplicated()]

In [315]:
products_wrangled.shape

(49688, 5)

In [316]:
products.shape

(49693, 5)

## EXPORTING "products_wrangled DATAFRAME

In [317]:
#DEFINE PATH
path = r'/Users/MadrigalPaez/Documents/2 - FORMACIÓN Y TRABAJO/CAREER FOUNDRY/DATA ANALYTICS/2 - Data Immersion/Achievement 4 - Python Fundamentals for Data Analysts/08-2025 Instacart Grocery Basket Analysis'

In [318]:
products_wrangled.to_csv(os.path.join(path, '02 Data','Prepared Data', 'products_wrangled.csv'))

# 4.5: Task

## If you haven’t performed the consistency checks covered in this Exercise on your df_prods dataframe, do so now:

Done in line [#100]

## Run the df.describe() function on your df_ords dataframe. Using your new knowledge about how to interpret the output of this function, share in a markdown cell whether anything about the data looks off or should be investigated further:

In [319]:
orders.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       order_day_of_week  order_hour_of_day  days_since_prior_order  
count       3.421083e+06       3.421083e+06            3.214874e+06  
mean        2.776219e+00       1.345202e+01            1.111484e+01  
std         2.046829e+00       4.226088e+00            9.206737e+00  
min         0.000000e+00       0.000000e+00            0.000000e+00  
25%         1.000000e+00       1.000000e+01            4.000000e+00  
50%         3.000000e+00       1.300000e+01            7.000000e+00  
75%         5.000000e+00       1.600000e+01            1.500000e+01  
max         6.000000e+00       2.300000e+01            3.000000e+01

In [320]:
#Changing numbers' format:
orders.describe().map(lambda x: f'{x:,.0f}')

Unnamed: 0   order_id    user_id order_number order_day_of_week  \
count  3,421,083  3,421,083  3,421,083    3,421,083         3,421,083   
mean   1,710,541  1,710,542    102,978           17                 3   
std      987,582    987,582     59,534           18                 2   
min            0          1          1            1                 0   
25%      855,270    855,272     51,394            5                 1   
50%    1,710,541  1,710,542    102,689           11                 3   
75%    2,565,812  2,565,812    154,385           23                 5   
max    3,421,082  3,421,083    206,209          100                 6   

      order_hour_of_day days_since_prior_order  
count         3,421,083              3,214,874  
mean                 13                     11  
std                   4                      9  
min                   0                      0  
25%                  10                      4  
50%                  13                      7  
75%                  16                     15  
max                  23                     30

- order_id: From 1 as Min, to 3,421,083 as Max. One "user_id" identifier per row.
- user_id: From 1 as Min, to 206,209 as Max, meaning the total individual users.
- order_number: Max is 100. 
- order_day_of_week: Max is 6, min is 0. One number for each day of the week.
- order_hour_of_day: Max is 24, so it uses the 24-hour format, from 0 to 23.
- days_since_prior_order: Max difference in terms of days between orders in this dataframe is 30 days.

## Check for mixed-type data in your df_ords dataframe:

In [321]:
inconsistent_rows = {}
for col in orders.columns:
    common_type = orders[col].map(type).mode()[0]
    weird = orders[col].map(type) != common_type
    if weird.any():
        inconsistent_rows[col] = products.loc[weird, [col]]
        print(f"{col}: {weird.sum()}")
    else:
        print(f"{col}: No inconsistencies found")

Unnamed: 0: No inconsistencies found
order_id: No inconsistencies found
user_id: No inconsistencies found
eval_set: No inconsistencies found
order_number: No inconsistencies found
order_day_of_week: No inconsistencies found
order_hour_of_day: No inconsistencies found
days_since_prior_order: No inconsistencies found


## Run a check for missing values in your df_ords dataframe. In markdown cell, report your findings and propose an explanation for any missing values you find: 

In [322]:
orders.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
eval_set                       0
order_number                   0
order_day_of_week              0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

"days_since_prior_order" contains 206,209 missing values. As we observed earlier, the maximum interval between orders in this dataset is 30 days.

## Address the missing values using an appropriate method. In a markdown cell, explain why you used your method of choice:

To approach this, we need to take into consideration two different scenarios:

- According to orders.describe(), the Max interval between orders is 30 days. This suggests that the dataset only records "days_since_prior_order" up to 30 days.
- OR
- All 206,209 missing values correspond to the users first order, since there is no prior order to calculate the difference from.

In [323]:
# Testing the first order equal NaN for days_since_prior_order hypothesis:
first_order = orders[orders['days_since_prior_order'].isna()]
first_order['order_number'].value_counts()

order_number
1    206209
Name: count, dtype: int64

All 206209 missing values correspond to "order_number" 1. That could probe that when "days_since_prior_order" = NaN is the first order. But we have also rows with "days_since_prior_order" = 0. Therefore we have again two different scenarios:

- "days_since_prior_order" = 0 means always first order.
- "days_since_prior_order" = 0 includes orders made on the same day.

In [324]:
# Testing the "days_since_prior_order" = 0 includes orders made on the same day using the order number:
orders[orders['days_since_prior_order'] == 0]['order_number'].value_counts()

order_number
4      2964
3      2915
2      2899
5      2634
6      2443
       ... 
98      144
100     137
99      134
93      125
97      120
Name: count, Length: 99, dtype: int64

Therefore: 0 = repeat order on the same day, >0 = number of days since the previous order. We can asume that NaN = first order and change it

In [325]:
orders[orders['days_since_prior_order'] == 0].shape[0]

67755

In [326]:
orders[orders['days_since_prior_order'] == 1].shape[1]

8

In [327]:
orders['days_since_prior_order'].dtypes

dtype('float64')

I wouldn't replace the NaN in "days_since_prior_order" with "First Order", as this would convert a numeric column into a string one. This will interfere with future calculations. Instead, we can create a new integer column named "first_order", where 0 = Not first order and 1 = First order. This preserves the numeric column for calculations, and provides a clear indicator for first orders.

In [328]:
orders.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
eval_set                       0
order_number                   0
order_day_of_week              0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

In [329]:
# Create new column "first_order" where "0 = Not first order" and "1 = First order"
orders['first_order'] = orders['days_since_prior_order'].isna().astype(int)

In [330]:
orders.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
eval_set                       0
order_number                   0
order_day_of_week              0
order_hour_of_day              0
days_since_prior_order    206209
first_order                    0
dtype: int64

In [331]:
orders['first_order'].value_counts()

first_order
0    3214874
1     206209
Name: count, dtype: int64

Now we have a new column where we can see the first orders. "days_since_prior_order" remains as it was before for analysis. The number of NaN we have on "days_since_prior_order" matches the number of "1" we have on "first_order".

## Run a check for duplicate values in your df_ords data. In a markdown cell, report your findings and propose an explanation for any duplicate values you find:

In [332]:
#Finding duplicates on "orders" dataframe:
orders_duplicates = orders[orders.duplicated()]

In [333]:
orders_duplicates

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, eval_set, order_number, order_day_of_week, order_hour_of_day, days_since_prior_order, first_order]
Index: []

No duplicate values have been found.

## Export your final, cleaned df_prods and df_ords data as “.csv” files in your “Prepared Data” folder and give them appropriate, succinct names:

In [2]:
#ESTABLISHING PATH
path = r'/Users/MadrigalPaez/Documents/2 - FORMACIÓN Y TRABAJO/CAREER FOUNDRY/DATA ANALYTICS/2 - Data Immersion/Achievement 4 - Python Fundamentals for Data Analysts/08-2025 Instacart Grocery Basket Analysis'

In [1]:
orders.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_cleaned.csv'))

NameError: name 'orders' is not defined

## When too many users make Instacart orders at the same time, the app freezes. The senior technical officer at Instacart wants you to identify the busiest hours of the day. 5.- Rather than by hour, they want periods of time labeled “Most orders,” “Average orders,” and “Fewest orders.” Create a new column containing these labels called “busiest_period_of_day.”:
